### PR TITLE
4847239: (spec) File.createTempFile() should make it clear that it doesn't create the temporary directory

### PR DESCRIPTION
**Pull request by bplb**
Augment the specification of `java.io.File.createTempFile(String,String,File)` to clarify its behavior with respect to the `File` parameter `directory`.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-4847239](https://bugs.openjdk.java.net/browse/JDK-4847239): (spec) File.createTempFile() should make it clear that it doesn't create the temporary directory

### Reviewers
 * [Naoto Sato](https://openjdk.java.net/census#naoto) (@naotoj - **Reviewer**)
 * [Lance Andersen](https://openjdk.java.net/census#lancea) (@LanceAndersen - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4561/head:pull/4561` \
`$ git checkout pull/4561`

Update a local copy of the PR: \
`$ git checkout pull/4561` \
`$ git pull https://git.openjdk.java.net/jdk pull/4561/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4561`

View PR using the GUI difftool: \
`$ git pr show -t 4561`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4561.diff">https://git.openjdk.java.net/jdk/pull/4561.diff</a>

</details>

### CODE REVIEW

**Review by naotoj — APPROVED** [review]
Looks good. Do we have a test case that verifies the behavior?

**Comment by bplb** [comment]
Surprisingly it does not look like there is a verifying test. I checked it manually but I had better add a specific test.

**Comment by bplb** [comment]
Test added.

File: src/java.base/share/classes/java/io/File.java [hunk 1]
```diff
@@ -2110,6 +2110,13 @@ static File generateFile(String prefix, String suffix, File dir)
      * is invoked, but programmatic changes to this property are not guaranteed
      * to have any effect upon the temporary directory used by this method.
      *
+     * <p> If the {@code directory} argument is not {@code null}, then if its
+     * abstract pathname is valid and denotes an existing directory, then the
+     * file will be created in that directory; otherwise, the file will not
+     * be created and an {@code IOException} will be thrown.  In no case will
```
Comment (LanceAndersen) [inline src/java.base/share/classes/java/io/File.java:2116]: I might suggest  reworking the 1st sentence as it is a tad long and we could probably reduce the 'then' usages.

**Review by naotoj — COMMENTED** [review]
Thanks for adding the test. Some nits follow.

File: test/jdk/java/io/File/createTempFile/TargetDirectory.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        File tmp = File.createTempFile("passes", null, target);
+        if (!Files.exists(tmp.toPath())) {
+            throw new RuntimeException("Temp file not created");
+        }
+        tmp.delete();
+
+        // Make target directory read-only
+        if (Files.getFileStore(dir).supportsFileAttributeView("posix")) {
+            PosixFileAttributeView view =
+                (PosixFileAttributeView)Files.getFileAttributeView(dir,
```
[the remark this answers is not in the document]
Comment (bplb) [reply test/jdk/java/io/File/createTempFile/TargetDirectory.java:58]: Yes, I think you are correct.

File: test/jdk/java/io/File/createTempFile/TargetDirectory.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            }
+            view.setAcl(entries);
+        } else {
+            throw new RuntimeException("Required attribute view not supported");
+        }
+
+        // Target directory exists but is read-only
+        try {
+            File.createTempFile("readonly", null, target);
+            throw new RuntimeException("Exception not thrown");
```
[the remark this answers is not in the document]
Comment (bplb) [reply test/jdk/java/io/File/createTempFile/TargetDirectory.java:83]: I was not sure a message was even needed as one can identify what is happening by the line number in the stack trace. Probably clear messages would be better though.

**Review by naotoj — APPROVED** [review]

**Review by LanceAndersen — APPROVED** [review]

**Pull request closed without merging**